src/cpu/dyn_cache.h

```diff
@@ -710,12 +723,50 @@ static void cache_block_closing(const uint8_t *block_start, Bitu block_size);
 #endif
 
 /* Define temporary pagesize so the MPROTECT case and the regular case share as much code as possible */
-#if defined(HAVE_MPROTECT)
+#if defined(HAVE_MPROTECT) || defined(HAVE_MMAP)
 #define PAGESIZE_TEMP PAGESIZE
 #else
 #define PAGESIZE_TEMP 4096
 #endif
 
+static constexpr size_t cache_code_start_sz = CACHE_TOTAL + CACHE_MAXSIZE + PAGESIZE_TEMP - 1 + PAGESIZE_TEMP;
+static constexpr size_t cache_blocks_sz = CACHE_BLOCKS * sizeof(CacheBlock);
+
+static inline void dyn_mem_execute()
+{
+#if defined(HAVE_APPLE_WRITEPROT)
+    if (__builtin_available(macOS 11.0, *)) {
+        pthread_jit_write_protect_np(true);
+    }
+#elif defined(HAVE_MMAP) && defined(HAVE_MPROTECT)
+	if (mprotect(cache_code_start_ptr, cache_code_start_sz, PROT_EXEC | PROT_READ) == -1) {
+		E_Exit("mprotect failed in dyn_mem_execute");		
+	}
+#endif
+}
+
+static inline void dyn_mem_write()
+{
+#if defined(HAVE_APPLE_WRITEPROT)
+    if (__builtin_available(macOS 11.0, *)) {
+        pthread_jit_write_protect_np(false);
+    }
+#elif defined(HAVE_MMAP) && defined(HAVE_MPROTECT)
+	if (mprotect(cache_code_start_ptr, cache_code_start_sz, PROT_WRITE | PROT_READ) == -1) {
+		E_Exit("mprotect failed in dyn_mem_write");			
```

Comment:
If we're only calling `mprotect` here, is `HAVE_MPROTECT` sufficient on it's own?

Reply:
Good question - I don't know enough about mmap and mprotect availability on different platforms. Could a platform have mprotect but not mmap?

Reply:
That shouldn't matter, yes, I'll remove the HAVE_MMAP check, since we'd still want to call the mprotect if there was no mmap and the memory was just plain malloc'd.

Reply:
Is there something about this call to `mprotect` that's using or (under the hood) expects to have `mmap` available too?  

I could state it like this: If  `HAVE_MPROTECT` is defined, then I should be able to confidently make a call to `mprotect` without other HAVE_'s needed. 

If that's not the case, the perhaps we need more checks up in Meson before handing out a `HAVE_MPROTECT` bus pass :-)

Reply:
> Could a platform have mprotect but not mmap?

Yes:
 - mprotect was first defined in the SVr4 spec (circa 1988), then carried into POSIX.1-2001.
 - map_shared and map_private were first defined in POSIX.1-2001 (roughly 2003)

Then again, I doubt anyone will be building Staging on a SVr4 (or greater) OS in those gap-years prior to ~2003 :sweat_smile: